File: src/banks/tangerine/schemas.ts
import { tz } from "@date-fns/tz";
import { formatISO } from "date-fns";
import { v5 as uuidv5 } from "uuid";
import { z } from "zod";
import env from "../../utils/env.js";

const Account = z
  .object({
    type: z.enum(["CHEQUING", "SAVINGS", "CREDIT_CARD"]),
    number: z.string(),
    account_balance: z.number(),
    display_name: z.string(),
    description: z.string(),
  })
  .transform((account) => ({
    id: uuidv5(account.number, env.UUID_NAMESPACE),
    name: `${account.description} (${account.display_name})`,
    balance:
      account.account_balance * (account.type === "CREDIT_CARD" ? -1 : 1),
    _number: account.number,
  }));

const AccountResponse = z
  .object({
    accounts: z.array(Account),
  })
  .transform(({ accounts }) => accounts);

const Transaction = z
  .object({
    transaction_date: z.string(),
    posted_date: z.string(),
    amount: z.number(),
    description: z.string(),
    is_uncleared: z.boolean(),
    status: z.string(),
  })
  .transform((transaction) => {
    const date =
      transaction.amount > 0
        ? transaction.posted_date
        : transaction.transaction_date;
    return {
      date: formatISO(date, {
        representation: "date",
        in: tz(env.TZ),
      }),
      amount: transaction.amount,
      description: transaction.description,
    };
  });

const TransactionsResponse = z
  .object({
    transactions: z.array(Transaction),
  })
  .transform(({ transactions }) => transactions);

export { Account, AccountResponse, TransactionsResponse };

File: src/banks/tangerine/schemas.py
from datetime import datetime, timezone, timedelta
from uuid import uuid5
from pydantic import BaseModel, Field, validator
import env

class Account(BaseModel):
    type: str
    number: str
    account_balance: float
    display_name: str
    description: str
    id: str = Field(default_factory=lambda: uuid5(env.UUID_NAMESPACE, self.number))
    name: str = Field(default_factory=lambda: f"{self.description} ({self.display_name})")
    balance: float = Field(default_factory=lambda: self.account_balance * (-1 if self.type == "CREDIT_CARD" else 1))
    _number: str = Field(default_factory=lambda: self.number)

class AccountResponse(BaseModel):
    accounts: list[Account]

class Transaction(BaseModel):
    transaction_date: str
    posted_date: str
    amount: float
    description: str
    is_uncleared: bool
    status: str
    date: str = Field(default_factory=lambda: format_iso(self.posted_date if self.amount > 0 else self.transaction_date))

    @validator('date', pre=True, always=True)
    def format_iso(cls, v):
        return datetime.fromisoformat(v).astimezone(timezone(timedelta(hours=int(env.TZ)))).isoformat()

class TransactionsResponse(BaseModel):
    transactions: list[Transaction]

class DisplayChallengeQuestionResponse(BaseModel):
    MessageBody: dict

    @validator('MessageBody', pre=True, always=True)
    def extract_question(cls, v):
        return v['Question']
